web/src/context/AtlasProvider.tsx

```diff
@@ -0,0 +1,316 @@
+import {
+  createMessage,
+  getNonce,
+  loginUser,
+  addUser as addUserToAtlas,
+  fetchUser,
+  updateEmail as updateEmailInAtlas,
+  confirmEmail as confirmEmailInAtlas,
+  uploadToIpfs,
+  type User,
+  type AddUserData,
+  type UpdateEmailData,
+  type ConfirmEmailData,
+  type ConfirmEmailResponse,
+  Roles,
+  Products,
+} from "utils/atlas";
+
+import { isUndefined } from "src/utils";
+
+interface IAtlasProvider {
+  isVerified: boolean;
+  isSigningIn: boolean;
+  isAddingUser: boolean;
+  isFetchingUser: boolean;
+  isUpdatingUser: boolean;
+  isUploadingFile: boolean;
+  user: User | undefined;
+  userExists: boolean;
+  authoriseUser: () => void;
+  addUser: (userSettings: AddUserData) => Promise<boolean>;
+  updateEmail: (userSettings: UpdateEmailData) => Promise<boolean>;
+  uploadFile: (file: File, role: Roles) => Promise<string | null>;
+  confirmEmail: (userSettings: ConfirmEmailData) => Promise<
+    ConfirmEmailResponse & {
+      isError: boolean;
+    }
+  >;
+}
+
+const Context = createContext<IAtlasProvider | undefined>(undefined);
+
+const atlasUri: string = import.meta.env.REACT_APP_ATLAS_URI ?? "";
+if (!atlasUri) {
+  console.warn("REACT_APP_ATLAS_URI is not defined. Please check your environment variables.");
+}
+
+const AtlasProvider: React.FC<{ children?: React.ReactNode }> = ({ children }) => {
+  const { address } = useAccount();
+  const chainId = useChainId();
+  const queryClient = useQueryClient();
+  const [authToken, setAuthToken] = useSessionStorage<string | undefined>("authToken", undefined);
+  const [isSigningIn, setIsSigningIn] = useState(false);
+  const [isAddingUser, setIsAddingUser] = useState(false);
+  const [isUpdatingUser, setIsUpdatingUser] = useState(false);
+  const [isVerified, setIsVerified] = useState(false);
+  const [isUploadingFile, setIsUploadingFile] = useState(false);
+  const { signMessageAsync } = useSignMessage();
+
+  const atlasGqlClient = useMemo(() => {
+    const headers = authToken
+      ? {
+          authorization: `Bearer ${authToken}`,
+        }
+      : undefined;
+    return new GraphQLClient(`${atlasUri}/graphql`, { headers });
+  }, [authToken]);
+
+  /**
+   * @description verifies user authorisation
+   * @returns boolean - true if user is authorized
+   */
+  const verifySession = useCallback(() => {
+    try {
+      if (!authToken || !address) return false;
+
+      const payload = decodeJwt(authToken);
+
+      if ((payload?.sub as string)?.toLowerCase() !== address.toLowerCase()) return false;
+      if (payload.exp && payload.exp < Date.now() / 1000) return false;
+
+      return true;
+    } catch {
+      return false;
+    }
+  }, [authToken, address]);
+
+  useEffect(() => {
+    // initial verfiy check
+    setIsVerified(verifySession());
+
+    // verify session every 5 sec
+    const intervalId = setInterval(() => {
+      setIsVerified(verifySession());
+    }, 5000);
+
+    return () => {
+      clearInterval(intervalId);
+    };
```

Comment:
Please create a new issue about this

Reply:
Hello,  i tought about a dynamic approach for the setInterval ie: considering site usage behavior, mixed that concept on geminiAI with the coderabbitai presented refactor example code. This come up 

## Combining the Best of Both Worlds: A Hybrid Approach

**Understanding the Strengths:**

- **Dynamic Interval Approach:** Offers flexibility in adjusting verification frequency based on session duration and user activity.
- **JWT Expiration Approach:** Leverages the built-in expiration mechanism of JWT tokens for precise verification scheduling.

**Hybrid Strategy:**

We can combine these approaches to create a more robust and efficient session verification strategy:

1. **Initial Verification:**
   - Use the dynamic interval approach to establish an initial verification interval.
   - This allows for immediate verification and flexibility in adjusting the interval based on early session behavior.

2. **JWT-Based Scheduling:**
   - Once a JWT token is acquired, extract its expiration time.
   - Schedule the next verification 1 minute before the expiration time, ensuring timely checks based on the token's validity.

3. **Dynamic Interval Adjustments:**
   - Continue to use the dynamic interval approach to adjust the verification interval based on user activity and session duration, even if a JWT-based schedule is in place.
   - This provides a safety net in case the JWT token becomes invalid or unexpected conditions arise.

**Implementation:**

```javascript
const useEffect = useCallback(() => {
  let verificationInterval = 30 * 1000; // Base interval (30 seconds)
  let sessionStartTime = Date.now();

  const verifyAndSchedule = () => {
    const isValid = verifySession();
    setIsVerified(isValid);

    if (isValid && authToken) {
      try {
        const payload = decodeJwt(authToken);
        const expiresIn = (payload.exp as number) * 1000 - Date.now();
        setTimeout(verifyAndSchedule, Math.max(0, expiresIn - 60000)); // Verify 1 minute before expiration
      } catch (err) {
        console.error("Error decoding JWT:", err);
        setIsVerified(false); // Set to false on decoding error
      }
    }

    // Adjust verification interval based on session duration
    const elapsedTime = Date.now() - sessionStartTime;
    verificationInterval = verificationInterval * (1 + Math.log(elapsedTime / 60000)); // Adjust the formula as needed
  };

  verifyAndSchedule();

  return () => {
    // Clear any scheduled verification
    clearTimeout(); // Assuming you have a reference stored for the timeout
  };
}, [authToken, verifySession]);
```

**Key Points:**

- This hybrid approach combines the precision of JWT-based scheduling with the flexibility of dynamic intervals.
- It provides a robust and efficient mechanism for session verification.
- You can adjust the base interval and the formula for dynamic interval adjustments to tailor the strategy to your specific needs.

By adopting this hybrid strategy, you can optimize session verification while maintaining a high level of security and performance.

Reply:
Hello, Thanks for taking time to suggest the strategy. Much appreciated.
We will be tackling this optimization in issue #1717 . You are welcome to review the applied strategy.

Thanks for the feedback 😊